src/System.IO.Pipelines/WritableBufferWriter.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.IO.Pipelines
+{
+    public struct WritableBufferWriter
+    {
+        private WritableBuffer _writableBuffer;
+        private Span<byte> _span;
+
+        public WritableBufferWriter(WritableBuffer writableBuffer)
+        {
+            _writableBuffer = writableBuffer;
+            _span = writableBuffer.Buffer.Span;
+        }
+
+        public Span<byte> Span => _span;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Advance(int count)
+        {
+            _span = _span.Slice(count);
+            _writableBuffer.Advance(count);
+        }
+
+        public void Write(byte[] source)
+        {
+            Write(source, 0, source.Length);
+        }
+
+        public void Write(byte[] source, int offset, int length)
+        {
+            if (length <= _span.Length && length > 0)
+            {
+                ref byte pSource = ref source[offset];
+                ref byte pDest = ref _span.DangerousGetPinnableReference();
+
+                Unsafe.CopyBlockUnaligned(ref pDest, ref pSource, (uint)length);
+
+                Advance(length);
+                return;
+            }
+
+            WriteMultiBuffer(source, offset, length);
+        }
+
+        private void WriteMultiBuffer(byte[] source, int offset, int length)
+        {
+            var remaining = length;
```

Comment:
Might not make much difference (not common path) and would need to look as asm; but wonder if arranging it like
```csharp
private void WriteMultiBuffer(byte[] source, int offset, int length)
{
    var finalOffset = offset + length;

    while (offset < finalOffset)
    {
        if (_span.Length == 0)
        {
            Ensure();
        }

        var writable = Math.Min(finalOffset - offset, _span.Length);

        Unsafe.CopyBlockUnaligned(ref _span.DangerousGetPinnableReference(), ref source[offset], (uint)writable);

        offset += writable;
        
        Advance(writable);
    }
}
```

Would do anything